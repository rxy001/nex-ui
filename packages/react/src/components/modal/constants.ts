export const MODAL_INTERNAL_ID_PREFIX = 'nui-modal-'
