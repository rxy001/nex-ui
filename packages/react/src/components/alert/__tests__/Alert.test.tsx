import {
  renderWithNexUIProvider,
  testClassNamesForwarding,
  testColorDataAttrs,
  testComponentStability,
  testRadiusDataAttrs,
  testRefForwarding,
  testRootClassName,
  testSlotPropsForwarding,
  testVariantDataAttrs,
} from '~/tests/shared'
import { Alert } from '../index'
import { alertSlotClasses, alertDataAttrs } from './constants'

const slots = [
  'icon',
  'content',
  'title',
  'description',
  'closeButton',
] as const

describe('Alert', () => {
  testComponentStability(<Alert />)

  testRootClassName(<Alert />)

  testColorDataAttrs(<Alert />)

  testRadiusDataAttrs(<Alert />)

  testVariantDataAttrs(<Alert />, [
    'status',
    ['error', 'success', 'warning', 'info'],
  ])

  testRefForwarding(<Alert />)

  testClassNamesForwarding(
    <Alert closable title='Title' description='Description' />,
    slots,
    {
      icon: 'test-icon',
      content: 'test-content',
      title: 'test-title',
      description: 'test-description',
      closeButton: 'test-close-button',
    },
    alertSlotClasses,
  )

  testSlotPropsForwarding(
    <Alert title='Title' description='Description' closable />,
    slots,
    {
      icon: {
        className: 'test-icon',
      },
      content: {
        className: 'test-content',
      },
      title: {
        className: 'test-title',
      },
      description: {
        className: 'test-description',
      },
      closeButton: {
        className: 'test-close-button',
      },
    },
    alertSlotClasses,
  )

  it('should render with default props', () => {
    const { container } = renderWithNexUIProvider(<Alert />)

    const alertRoot = container.firstChild

    expect(alertRoot).toHaveClass(alertSlotClasses.root)

    expect(alertRoot).toHaveAttribute(...alertDataAttrs['status-info'])
    expect(alertRoot).toHaveAttribute(...alertDataAttrs['variant-faded'])
    expect(alertRoot).toHaveAttribute(...alertDataAttrs['radius-md'])
    expect(alertRoot).toHaveAttribute(...alertDataAttrs['color-blue'])

    expect(alertRoot).toMatchSnapshot()
  })

  it('should render close button when closable is true', () => {
    const { queryByClassName, rerender } = renderWithNexUIProvider(<Alert />)

    expect(
      queryByClassName(alertSlotClasses['close-button']),
    ).not.toBeInTheDocument()

    rerender(<Alert closable />)

    expect(
      queryByClassName(alertSlotClasses['close-button']),
    ).toBeInTheDocument()
  })

  it('should hide icon when hideIcon is true', () => {
    const { queryByClassName, rerender } = renderWithNexUIProvider(<Alert />)

    expect(queryByClassName(alertSlotClasses.icon)).toBeInTheDocument()

    rerender(<Alert hideIcon />)

    expect(queryByClassName(alertSlotClasses.icon)).not.toBeInTheDocument()
  })

  it('should override status color when color is provided', () => {
    const { queryByClassName, rerender } = renderWithNexUIProvider(
      <Alert color='purple' status='info' />,
    )

    const alertRoot = queryByClassName(alertSlotClasses.root)

    expect(alertRoot).toHaveAttribute(...alertDataAttrs['color-purple'])

    rerender(<Alert color='green' status='success' />)

    expect(alertRoot).toHaveAttribute(...alertDataAttrs['color-green'])
  })

  it('should custom icon', () => {
    const { queryByTestId } = renderWithNexUIProvider(
      <Alert icon={<span data-testid='custom-icon' />} />,
    )

    expect(queryByTestId('custom-icon')).toBeInTheDocument()
  })

  it('should custom action', () => {
    const { queryByTestId } = renderWithNexUIProvider(
      <Alert action={<span data-testid='custom-action' />} />,
    )

    expect(queryByTestId('custom-action')).toBeInTheDocument()
  })

  it('should render description slot when description is provided', () => {
    const { queryByClassName, rerender } = renderWithNexUIProvider(<Alert />)

    expect(
      queryByClassName(alertSlotClasses.description),
    ).not.toBeInTheDocument()

    rerender(<Alert description='Description' />)

    expect(queryByClassName(alertSlotClasses.description)).toBeInTheDocument()
  })

  it('should error when status is invalid', () => {
    const consoleSpy = jest.spyOn(console, 'error').mockImplementation()

    // @ts-expect-error
    renderWithNexUIProvider(<Alert status='unknown' />)

    expect(consoleSpy).toHaveBeenCalledWith(
      '[Nex UI] Alert: Unknown status %s',
      'unknown',
    )

    consoleSpy.mockRestore()
  })

  it('should call onClose when close button is clicked', async () => {
    const onClose = jest.fn()
    const { queryByClassName, user } = renderWithNexUIProvider(
      <Alert closable onClose={onClose} />,
    )
    await user.click(queryByClassName(alertSlotClasses['close-button'])!)
    expect(onClose).toHaveBeenCalledTimes(1)
  })

  describe('Accessibility', () => {
    it('should have role alert on root element', () => {
      const { container } = renderWithNexUIProvider(<Alert />)

      expect(container.firstElementChild).toHaveAttribute('role', 'alert')
    })

    it('should have aria-label on close button', () => {
      const { queryByClassName } = renderWithNexUIProvider(<Alert closable />)

      expect(
        queryByClassName(alertSlotClasses['close-button']),
      ).toHaveAttribute('aria-label', 'Close alert')
    })
  })
})
