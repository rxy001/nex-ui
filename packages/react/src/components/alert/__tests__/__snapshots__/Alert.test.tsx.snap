// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert should render with default props 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-padding-start: var(--nui-spaces-4);
  padding-inline-start: var(--nui-spaces-4);
  -webkit-padding-end: var(--nui-spaces-4);
  padding-inline-end: var(--nui-spaces-4);
  padding-block-start: var(--nui-spaces-3);
  padding-block-end: var(--nui-spaces-3);
  gap: var(--nui-spaces-3);
  overflow: hidden;
  background-color: var(--nui-colors-blue-subtle);
  color: var(--nui-colors-blue-600);
  border-radius: var(--nui-radii-lg);
}

[data-nui-color-scheme="dark"] .emotion-0 {
  color: var(--nui-colors-blue-400);
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  font-size: 26px;
}

.emotion-2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  min-height: var(--nui-sizes-8);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: var(--nui-spaces-1);
  color: inherit;
}

<div
  class="nui-alert-root emotion-0"
  data-color="blue"
  data-radius="md"
  data-status="info"
  data-variant="faded"
  role="alert"
>
  <div
    class="nui-alert-icon emotion-1"
  >
    <svg
      aria-hidden="true"
      focusable="false"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M22 12c0-5.523-4.477-10-10-10S2 6.477 2 12s4.477 10 10 10s10-4.477 10-10M12 6.25a.75.75 0 0 1 .75.75v6a.75.75 0 0 1-1.5 0V7a.75.75 0 0 1 .75-.75M12 17a1 1 0 1 0 0-2a1 1 0 0 0 0 2"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
  </div>
  <div
    class="nui-alert-content emotion-2"
  />
</div>
`;
